src/main/java/org/broadinstitute/hellbender/tools/walkers/fasta/ShiftFasta.java

```diff
@@ -0,0 +1,210 @@
+    private String intervalFilename;
+
+    public static final String LINE_WIDTH_LONG_NAME = "line-width";
+    @Argument(fullName= LINE_WIDTH_LONG_NAME, doc="Maximum length of sequence to write per line", optional=true)
+    public int basesPerLine = FastaReferenceWriter.DEFAULT_BASES_PER_LINE;
+
+    ReferenceDataSource refSource;
+    FastaReferenceWriter refWriter;
+    FileWriter chainFileWriter;
+    FileWriter intervalRegularWriter;
+    FileWriter intervalShiftedWriter;
+
+    int chainId = 0;
+
+    @Override
+    public boolean requiresReference() {
+        return true;
+    }
+
+    @Override
+    public void onTraversalStart() {
+        refSource = referenceArguments.getReferencePath() != null ? ReferenceDataSource.of(referenceArguments.getReferencePath()) : null;
+        final Path path = IOUtils.getPath(output);
+        chainId = 1;
+        try {
+            refWriter = new FastaReferenceWriterBuilder()
+                    .setFastaFile(path)
+                    .setBasesPerLine(basesPerLine)
+                    .build();
+            chainFileWriter = new FileWriter(shiftBackOutput);
+            if (intervalFilename != null) {
+                intervalRegularWriter = new FileWriter(intervalFilename+ ".intervals");
+                intervalShiftedWriter = new FileWriter(intervalFilename + ".shifted.intervals");
+            }
+        } catch (IOException e) {
+            throw new UserException.CouldNotCreateOutputFile("Couldn't create " + output + ", encountered exception: " + e.getMessage(), e);
+        }
+    }
+
+    public void traverse() {
+        SAMSequenceDictionary refDict = refSource.getSequenceDictionary();
+        long refLengthLong = refDict.getReferenceLength();
+        if (refLengthLong > Integer.MAX_VALUE) {
+            // TODO fix this??
+            throw new UserException.BadInput("Reference length is too long");
+        }
+        List<SAMSequenceRecord> contigs = refSource.getSequenceDictionary().getSequences();
+        final ListIterator<Integer> shiftOffsetsIt = (shiftOffsets != null && shiftOffsets.size() == contigs.size()) ?
+                shiftOffsets.listIterator() : null;
+        refSource.getSequenceDictionary().getSequences().forEach(seq -> shiftContig(seq, shiftOffsetsIt));
+    }
+
+    protected void shiftContig(SAMSequenceRecord seq, ListIterator<Integer> shiftOffsetsIt) {
+        int contigLength = seq.getSequenceLength();
+        int shiftOffset = shiftOffsetsIt == null ? contigLength/2 : shiftOffsetsIt.next();
+        if (shiftOffset != 0) {
+            byte[] bases = refSource.queryAndPrefetch(new SimpleInterval(seq.getSequenceName(), 1, contigLength)).getBases();
+            byte[] basesAtEnd = Arrays.copyOfRange(bases, shiftOffset, bases.length);
+            byte[] basesAtStart = Arrays.copyOf(bases, shiftOffset);
+            int shiftBackOffset = bases.length - shiftOffset;
```

Comment:
Here you need to add some tests that assert the offset is within the range of the actual file? There seems to be no check that the user cant specify a bogus value in the list input.